plugin/client_rpc.go

```diff
@@ -0,0 +1,331 @@
+	impl API
+}
+
+// Registering some types used by MM for encoding/gob used by rpc
+func init() {
+	gob.Register([]*model.SlackAttachment{})
+	gob.Register([]interface{}{})
+	gob.Register(map[string]interface{}{})
+}
+
+// These enforce compile time checks to make sure types implemnt the interface
+// If your getting an error here, you probably need to run `make pluginapi` to
+// autogenerate RPC glue code
+var _ plugin.Plugin = &HooksPlugin{}
+var _ Hooks = &HooksRPCClient{}
+
+//
+// Below are specal cases for hooks or APIs that can not be auto generated
+//
+
+func (g *HooksRPCClient) Implemented() (impl []string, err error) {
+	err = g.client.Call("Plugin.Implemented", struct{}{}, &impl)
+	for _, hookName := range impl {
+		if hookId, ok := HookNameToId[hookName]; ok {
+			g.implemented[hookId] = true
+		}
+	}
+	return
+}
+
+// Implemented replies with the names of the hooks that are implemented.
+func (s *HooksRPCServer) Implemented(args struct{}, reply *[]string) error {
+	ifaceType := reflect.TypeOf((*Hooks)(nil)).Elem()
+	implType := reflect.TypeOf(s.impl)
+	selfType := reflect.TypeOf(s)
+	var methods []string
+	for i := 0; i < ifaceType.NumMethod(); i++ {
+		method := ifaceType.Method(i)
+		if m, ok := implType.MethodByName(method.Name); !ok {
+			continue
+		} else if m.Type.NumIn() != method.Type.NumIn()+1 {
+			continue
+		} else if m.Type.NumOut() != method.Type.NumOut() {
+			continue
+		} else {
+			match := true
+			for j := 0; j < method.Type.NumIn(); j++ {
+				if m.Type.In(j+1) != method.Type.In(j) {
+					match = false
+					break
+				}
+			}
+			for j := 0; j < method.Type.NumOut(); j++ {
+				if m.Type.Out(j) != method.Type.Out(j) {
+					match = false
+					break
+				}
+			}
+			if !match {
+				continue
+			}
+		}
+		if _, ok := selfType.MethodByName(method.Name); !ok {
+			continue
+		}
+		methods = append(methods, method.Name)
+	}
+	*reply = methods
+	return nil
+}
+
+type OnActivateArgs struct {
+	APIMuxId uint32
+}
+
+type OnActivateReturns struct {
+	A error
+}
+
+func (g *HooksRPCClient) OnActivate() error {
+	muxId := g.muxBroker.NextId()
+	go g.muxBroker.AcceptAndServe(muxId, &APIRPCServer{
+		impl: g.apiImpl,
+	})
+
+	_args := &OnActivateArgs{
+		APIMuxId: muxId,
+	}
+	_returns := &OnActivateReturns{}
+
+	if err := g.client.Call("Plugin.OnActivate", _args, _returns); err != nil {
+		g.log.Error("RPC call to OnActivate plugin failed.", mlog.Err(err))
+	}
+	return _returns.A
+}
+
+func (s *HooksRPCServer) OnActivate(args *OnActivateArgs, returns *OnActivateReturns) error {
+	connection, err := s.muxBroker.Dial(args.APIMuxId)
+	if err != nil {
+		return err // Where does this go?
+	}
+
+	// Settings for this should come from the parent process, for now just set it up
+	// though stdout.
+	logger := mlog.NewLogger(&mlog.LoggerConfiguration{
+		EnableConsole: true,
+		ConsoleJson:   true,
+		ConsoleLevel:  mlog.LevelDebug,
+		EnableFile:    false,
+	})
+	logger = logger.With(mlog.String("plugin_subprocess", ""))
```

Comment:
I guess this is a sort of tag? Would it make sense to log this as a boolean for eventual filtering vs. having to detect the presence of the key alone?